Simplify pipeline flow + add opt-in tool parameters

nvnieuwk opened the pull request:
This PR contains a refactor of the pipeline code to make it easier to read and understand how it works. 

The main changes are:
1. Added a `tools` input in the main workflow that contains all tools that need to be run (it will contain all tools when `--all` is used)
2. Simplified the `build_references` logic to make it less verbose and more stable. I also added tool specific checks so unused references are not created for the pipeline run
3. Removed some weird logic (like the `stub_check` parameters and empty dummy fusion files)
4. Removed `ext.when` for compatibility with future Nextflow versions

(I probably forgot to add some things here)

Some changes will have to be done if #633 is merged, but I'll take care of that once that's the case

Please check this PR very carefully and let me know if I did something stupid :grin:

rannick commented:
Yes I would like to test that one extensively. Will start some runs today

nvnieuwk commented:
> Yes I would like to test that one extensively. Will start some runs today

Thanks, let me know if you encounter any issues. I've done several tests with combination of parameters so most issues should be solved now. But extra testing won't hurt anyone indeed

atrigila reviewed (commented):
Thanks a lot for addressing all these issues, really nice work, this is a great contribution!

I did a test run with `nextflow run . -profile test_build,docker --outdir test_build` and it works as expected. I could even remove the `--references_only flag` to be able to run some of the main processes such as fusioninspector. So we are closer to having a full test that covers more processes and not only the building reference aspect :)

As a thought for future development: would it make sense to consolidate tool-specific flags into a single `--tools` parameter where users can list the tools they want to run? That might simplify things a bit.

No need to address any of this in this PR if you don't want to, just some ideas we can revisit for future design discussions.

nvnieuwk commented:
> Thanks a lot for addressing all these issues, really nice work, this is a great contribution!
> 
> I did a test run with `nextflow run . -profile test_build,docker --outdir test_build` and it works as expected. I could even remove the `--references_only flag` to be able to run some of the main processes such as fusioninspector. So we are closer to having a full test that covers more processes and not only the building reference aspect :)
> 
> As a thought for future development: would it make sense to consolidate tool-specific flags into a single `--tools` parameter where users can list the tools they want to run? That might simplify things a bit.
> 
> No need to address any of this in this PR if you don't want to, just some ideas we can revisit for future design discussions.

Awesome, thank you for these kind words! I indeed considered changing all tool parameters to `--tools` but didn't want to change that much here, but seeing that you are behind that idea made me change my mind. I'll update that now!

nvnieuwk commented on CHANGELOG.md:
```diff
@@ -51,6 +53,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Updated documentation for fusion-report score calculation to reflect 80/20 weight distribution between tool detection and database hits [#620](https://github.com/nf-core/rnafusion/pull/620)
 - Update htslib and samtools version in `star/align` to 1.21 [#634](https://github.com/nf-core/rnafusion/pull/634)
 - `--run_fusioncatcher` back to `fusioncatcher` [#641](https://github.com/nf-core/rnafusion/pull/641)
+- Removed the `--fusioninspector_only` parameter, this is now the case when only `--fusionreport` is used together with `--skip_qc`, `--skip_vis` and `--skip_vcf` [#645](https://github.com/nf-core/rnafusion/pull/645)
```
Yeah I fully agree that the `skip_` parameters should also be split up into their separate tools params or added to the `--tools` param somehow, but didn't want to touch this that much in this PR since it's already huge :grin:

rannick replied:
I agree that we could make the skipping cleaner, I think it already looks a lot better with the opt-in tools

nvnieuwk commented on conf/modules.config:
```diff
@@ -36,7 +36,6 @@ process {
     }
 
     withName: 'ARRIBA_VISUALISATION' {
-        ext.when   = { {!params.fusioninspector_only} && ({params.starfusion} || {params.all}) }
```
I noticed that most work for that was already done actually, it took me 5 minutes to fix up the couple exceptions

nvnieuwk commented on subworkflows/local/build_references.nf:
```diff
@@ -48,102 +52,121 @@ workflow BUILD_REFERENCES {
-        ch_hgnc_date = Channel.fromPath(params.hgnc_date).map { that -> [[id:that.Name], that] }
+    def ch_hgnc_date = Channel.empty()
+    def ch_hgnc_ref  = Channel.empty()
+    def run_fusioninspector = tools.contains("fusioninspector")
+    if(run_fusioninspector) {
```
Yes absolutely, this is sadly giving an error due to a false positive compile bug in Nextflow which should be fixed in the future. (I can't find the specific issue anymore), but you will see the error once you merge these into one line :)

nvnieuwk commented:
I'm also wondering what the `test_cosmic` test is actually trying to do. It seems like nothing actually runs in this test

atrigila commented on tests/test_build.nf.test:
```diff
@@ -11,7 +11,8 @@ nextflow_pipeline {
             params {
-                outdir               = "$outputDir"
+                outdir = "$outputDir"
+                tools  = "salmon,fusioninspector"
```
I think the test might be failing due to a mismatch between the tools specified here and those defined in the profile.

I'd suggest not adding tool parameters directly in the nf-test file, since it relies on the `test_build` profile. The tools used should be defined within that profile, otherwise, the params here would override the profile settings.

The `test_build.config` lists all the processes being tested for traceability, so keeping everything aligned there might help with maintainability.

nvnieuwk replied:
Thanks I didn't realize there was a separate config for that test :) I'll update it!

atrigila commented:
> Would you like to have tests for different tool combinations in place? Or should I keep that for another PR?

That's exactly what I was using to test this PR on my side! If you'd like to go ahead and add those tests, that would be great, but please don't feel any pressure to do more changes. What you've done already is great, and from my side, this PR is approved :)

The goal is that these CI tests will help speed up reviews. I used them to test the single STAR run PR (which worked well, by the way). The idea is to eventually have at least one CI test with all or some combination of the tools used, `fusioncatcher` might be the only tricky one for CI, but the rest should be doable. I won’t be able to submit PRs to this pipeline in the short term, but I’m definitely happy to review.

nvnieuwk commented:
> That's exactly what I was using to test this PR on my side! If you'd like to go ahead and add those tests, that would be great, but please don't feel any pressure to do more changes. What you've done already is great, and from my side, this PR is approved :)
> 
> The goal is that these CI tests will help speed up reviews. I used them to test the single STAR run PR (which worked well, by the way). The idea is to eventually have at least one CI test with all or some combination of the tools used, `fusioncatcher` might be the only tricky one for CI, but the rest should be doable. I won’t be able to submit PRs to this pipeline in the short term, but I’m definitely happy to review.

No problem, I'll add them today :) You are right that this will make future development a lot easier!

nvnieuwk commented:
I added tests for all separate tools, it wouldn't really make sense to test combination as most tools don't interact that much and they should already work properly if they would interact when only the single tool is used

rannick commented on subworkflows/local/utils_nfcore_rnafusion_pipeline/main.nf:
```diff
@@ -156,16 +156,7 @@ def validateInputParameters() {
-            log.warn("Skipping COSMIC DB download from `FUSIONREPORT_DOWNLOAD` and skip using it in `FUSIONREPORT`")
-    }
-
-    if (params.starfusion_build && !params.fusion_annot_lib) {
```
Why removing this warning?

nvnieuwk replied:
Because that parameter wasn't used anymore

nvnieuwk replied:
Would you like me to reimplement it?

rannick replied:
No, i just checked that fusion_annot_lib has "exists": true, but could we have a check for fusion_cannot_lib in build_references when building STARFUSION?

rannick replied:
Just to make sure it is defined?

nvnieuwk replied:
Good idea!

nvnieuwk replied:
I added the extra check :)

rannick commented on main.nf:
```diff
@@ -74,10 +39,28 @@ workflow {
     )
 
+    def tools = params.tools.tokenize(",")
+    if (tools.contains("all")) {
```
Trimming was not done by default, it is not a bad idea to include it, but documentation needs updating

nvnieuwk replied:
Done! Thanks :)

rannick commented on workflows/rnafusion.nf:
```diff
@@ -45,136 +48,215 @@ workflow RNAFUSION {
+    BUILD_REFERENCES(tools)
     ch_versions = ch_versions.mix(BUILD_REFERENCES.out.versions)
 
     if (!params.references_only) { // TODO: Remove this temporary parameter when we have a full-working GitHub nf-test
```
```suggestion
    if (!params.references_only) {
```

rannick replied:
this is not a temporary parameter but one that allows user to only build the references. As it takes ages and for documentation purposes, this behavior is wished. 

(not related to this PR but since I saw the comment, why not correct it 😄  )

nvnieuwk replied:
Fair enough, thanks ;)

rannick commented on workflows/rnafusion.nf:
```diff
@@ -45,136 +48,215 @@ workflow RNAFUSION {
+            ch_reads = ch_samplesheet
+        }
+
+        if(tools.contains("salmon")) {
```
```suggestion
        if(tools.contains("salmon") && !params.skip_qc) {
```

rannick commented on workflows/rnafusion.nf:
```diff
@@ -45,136 +48,215 @@ workflow RNAFUSION {
-        )
-        ch_versions = ch_versions.mix(ARRIBA_WORKFLOW.out.versions)
+        def fusions_created = false
+        def ch_arriba_fusions = ch_reads.map { it -> [it[0], []] } // Set arriba fusions to empty by default
```
Just for my curiosity, why not a Channel.empty() here?

rannick commented on subworkflows/local/fusioninspector_workflow.nf:
```diff
@@ -17,43 +17,49 @@ workflow FUSIONINSPECTOR_WORKFLOW {
 
-        AGAT_CONVERTSPGFF2TSV(FUSIONINSPECTOR.out.out_gtf)
-        ch_versions = ch_versions.mix(AGAT_CONVERTSPGFF2TSV.out.versions)
+        if(!skip_vcf) {
```
Feel free to ignore, but I am wondering if it would make sense to move this to the main workflow for logic readability

nvnieuwk replied:
I'll leave that up to you if that's okay, I'm not that familiar with all the different tools so your judgement on this is way better than mine

rannick replied:
Let's leave it there for the moment and see if it "flows" when we try to make new changes

rannick reviewed (commented):
Fantastic work! I did some nitpicking so I can fully understand what is going on and maybe avoid some issues downstream hopefully

rannick commented on subworkflows/local/build_references.nf:
```diff
@@ -48,102 +52,121 @@ workflow BUILD_REFERENCES {
+    def ch_arriba_ref_protein_domains = Channel.empty()
+    def arriba_tools = tools.intersect(["arriba", "fusioninspector"])
+    if (arriba_tools) {
+        if (!exists_not_empty(params.arriba_ref_blacklist) || !exists_not_empty(params.arriba_ref_known_fusions) || !exists_not_empty(params.arriba_ref_protein_domains)) {
```
(can be addressed in a later PR, but if you chose to ignore, could you open an issue so we track this please?)

nvnieuwk commented on subworkflows/local/build_references.nf:
```diff
@@ -48,102 +52,121 @@ workflow BUILD_REFERENCES {
-        ch_salmon_index = Channel.fromPath({params.salmon_index})
+    def ch_salmon_index = Channel.empty()
+    if (tools.contains("salmon")) {
+        if (!exists_not_empty(params.salmon_index)){ // add condition for qc
+            GFFREAD(ch_gtf, ch_fasta.map{ it -> it[1] })
+            ch_versions = ch_versions.mix(GFFREAD.out.versions)
+            SALMON_INDEX(ch_fasta.map{ it -> it[1] }, GFFREAD.out.gffread_fasta.map{ it -> it[1] })
+            ch_versions = ch_versions.mix(SALMON_INDEX.out.versions)
+            ch_salmon_index = SALMON_INDEX.out.index
+        } else {
+            ch_salmon_index = Channel.fromPath(params.salmon_index).map { it -> [[id:it.name], it] }
+        }
```
Good to know, thanks!

nvnieuwk commented on subworkflows/local/build_references.nf:
```diff
@@ -48,102 +52,121 @@ workflow BUILD_REFERENCES {
+    def ch_arriba_ref_cytobands       = Channel.empty()
+    def ch_arriba_ref_known_fusions   = Channel.empty()
+    def ch_arriba_ref_protein_domains = Channel.empty()
+    def arriba_tools = tools.intersect(["arriba", "fusioninspector"])
```
I'm not really sure, I might have made a small mistake here, I'll remove it :)

nvnieuwk commented on subworkflows/local/build_references.nf:
```diff
@@ -153,45 +176,53 @@ workflow BUILD_REFERENCES {
+def exists_not_empty(path) {
+    def path_to_check = file(path as String)
+    return path_to_check.exists() && !path_to_check.isEmpty()
+}
```
I updated the function, it now looks at the first file in the directory instead of at the directory directly

rannick replied:
Amazing@

nvnieuwk commented:
@rannick and @atrigila a massive thank you to both of you for your kind words and extensive reviews :). I addressed all your comments, is there something else missing before we can merge this PR?

rannick commented:
I think you can go ahead!

nvnieuwk commented:
Can you approve this PR? :)
